### PR TITLE
Add subscription validating method

### LINKED ISSUE #27: Control whether subscription is possible
Extend SubscriptionInterface:

    /**
     * @param Activity $activity
     * @throws GroupLife\Core\Exception\SubscriptionIsForbidden
     */
    public function assertValid(Activity $activity): void;

Implement this new method in each of implemented subscription objects. Utilize at `Activity::subscribe` method.

### PR DESCRIPTION
**Pull request by Lehersun**
- Add 'isValid' method to all  implemented subscription objects
- Add test with unvalid subsription

### CODE REVIEW

File: src/Subscription/Membership.php [hunk 1]
```diff
@@ -37,4 +39,11 @@ public function getPeriod(): \DateInterval
     {
         return $this->period;
     }
+
+    /**
+     * @param Core\Activity $activity
+     */
+    public function isValid(Core\Activity $activity): void
```
Comment (Magomogo) [inline src/Subscription/Membership.php]: My mistake, let's name it `assertValid`, if not it should throw.

File: src/Subscription/Activity.php [hunk 2]
```diff
@@ -45,4 +46,15 @@ public function getPeriod(): \DateInterval
...
         return $this->period;
     }
+
+    /**
+     * @param Core\Activity $activity
+     * @throws SubscriptionIsForbidden
+     */
+    public function isValid(Core\Activity $activity): void
+    {
+        if ($this->activity !== $activity) {
```
Comment (Magomogo) [inline src/Subscription/Activity.php]: You should not expect the same instance, let's use the activity name as an identifier.

File: src/Activity.php [hunk 3]
```diff
@@ -38,13 +39,24 @@ public function getLeader(): Leader
...
     {
         $activityVisits = [];
-        foreach ($this->schedule->materialize($subscription->getStartDay(), $subscription->getPeriod()) as $day) {
-            $activityVisits[] = new Visit($day, $this, $visitor);
+        try {
+            $subscription->isValid($this);
+            foreach ($this->schedule->materialize($subscription->getStartDay(), $subscription->getPeriod()) as $day) {
+                $activityVisits[] = new Visit($day, $this, $visitor);
+            }
+        } catch (SubscriptionIsForbidden $exception) {
```
Comment (Magomogo) [inline src/Activity.php]: Don't catch the exception here, let it be a part of business logic: we'll catch it in some client code.

**Review by Magomogo — CHANGES_REQUESTED** [review]

File: src/Activity.php (hunk 3 above)
Comment (Lehersun) [reply src/Activity.php]: Do I need to throw an exception forward? Otherwise, the subscription will be implemented

Comment (Lehersun) [reply src/Activity.php]: Or completely exclude the try-catch from this code block?

Comment (Magomogo) [reply src/Activity.php]: Yes, don't catch here anything.

**Pull request merged**